src/server/pkg/storage/chunk/writer.go

```diff
@@ -0,0 +1,174 @@
+package chunk
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"io"
+	"path"
+
+	"github.com/chmduquesne/rollinghash/buzhash64"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/hash"
+)
+
+const (
+	// MB is Megabytes.
+	MB = 1024 * 1024
+	// AverageBits determines the average chunk size (2^AverageBits).
+	AverageBits = 23
+	// WindowSize is the size of the rolling hash window.
+	WindowSize = 64
+)
+
+// Writer splits a byte stream into content defined chunks that are hashed and deduplicated/uploaded to object storage.
+// Chunk split points are determined by a bit pattern in a rolling hash function (buzhash64 at https://github.com/chmduquesne/rollinghash).
+// A min/max chunk size is also enforced to avoid creating chunks that are too small/big.
+// (bryce) The chunking/hashing/uploading could be made concurrent by reading ahead a certain amount and splitting the data among chunking/hashing/uploading workers
+// in a circular array where the first identified chunk (or whole chunk if there is no chunk split point) in a worker is appended to the prior workers data. This would
+// handle chunk splits that show up when the rolling hash window is across the data splits. The callback would still be executed sequentially so that the order would be
+// correct for the file index.
+// - An improvement to this would be to just append WindowSize bytes to the prior worker's data, then stitch together the correct chunks.
+//   It doesn't make sense to roll the window over the same data twice.
+// - This does not work with the Max criteria, need to revisit.
+type Writer struct {
+	ctx        context.Context
+	objC       obj.Client
+	prefix     string
+	cbs        []func([]*DataRef) error
+	buf        *bytes.Buffer
+	hash       *buzhash64.Buzhash64
+	splitMask  uint64
+	dataRefs   []*DataRef
+	done       [][]*DataRef
+	rangeSize  uint64
+	rangeCount uint64
+}
+
+// newWriter creates a new Writer.
+func newWriter(ctx context.Context, objC obj.Client, prefix string) *Writer {
+	// Initialize buzhash64 with WindowSize window.
+	hash := buzhash64.New()
+	hash.Write(make([]byte, WindowSize))
+	return &Writer{
+		ctx:       ctx,
+		objC:      objC,
+		prefix:    prefix,
+		cbs:       []func([]*DataRef) error{},
+		buf:       &bytes.Buffer{},
+		hash:      hash,
+		splitMask: (1 << uint64(AverageBits)) - 1,
+	}
+}
+
+// RangeStart specifies the start of a range within the byte stream that is meaningful to the caller.
+// When this range has ended (by calling RangeStart again or Close) and all of the necessary chunks are written, the
+// callback given during initialization will be called with DataRefs that can be used for accessing that range.
+func (w *Writer) RangeStart(cb func([]*DataRef) error) {
```

Comment:
How would you feel about calling this `StartRange`? I think it would make it a little more obvious that this is a method that has side effects, unlike `RangeSize` which just returns the size oft he range.

Reply:
Lol it was called `StartRange` originally, but I changed it to `RangeStart` because `SizeRange` did not make sense. I thought `RangeStart` made the most sense because it seems to be like qualifying an operation with Range. This would make all of the range operations be `Range<Operation>`. I am not too sure about this right now.

Reply:
So I was thinking that methods that methods that have side-effects would be name `VerbRange` and methods that just read would be caused `RangeNoun`. I don't know though, I don't feel too strongly about it.